[C10d] Add PG::enableCollectivesTiming to make it dynamically enabled.

kumpera opened the pull request:
Collectives timing gates the tracking when a collective starts on device.

Currently it's enabled by set the NCCL_ENABLE_TIMING env var.

The goal of this PR is to make it possible to dynamically enable that flag so users of the PG hooks don't have to set that flag in order to have their hooks work.

The design is that once set, all new collectives will have such behavior so we track it on each Work object.

We make enableTiming_ atomic in PGNCCL to avoid races on non-TSO hardware.

To ensure consistency, we copy its value during Work construction and replace all previous usage of enableTiming_ from the PG with usages from the Work, which now has an immutable value.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108815
* #109136
* __->__ #108814
* #108813

kumpera committed:
[C10d] Add PG::enableCollectivesTiming to make it dynamically enabled.

[ghstack-poisoned]

kumpera committed:
Update on "[C10d] Add PG::enableCollectivesTiming to make it dynamically enabled."




[ghstack-poisoned]

wconstab reviewed (commented):
a little confused at first, would appreciate a PR desc with more color about why.  IIUC you separated things so PG has 'enableTiming_' and each Work has 'timingEnabled_'.

Why is enableTiming_ atomic?

I assume the goal is to allow managing timing at the per-work scope, perhaps by toggling the PG flag before/after doing some comm ops?

wconstab reviewed (approved):
thanks for the Desc. makes sense to me.

kumpera committed:
Update on "[C10d] Add PG::enableCollectivesTiming to make it dynamically enabled."

Collectives timing gates the tracking when a collective starts on device.

Currently it's enabled by set the NCCL_ENABLE_TIMING env var.

The goal of this PR is to make it possible to dynamically enable that flag so users of the PG hooks don't have to set that flag in order to have their hooks work.

The design is that once set, all new collectives will have such behavior so we track it on each Work object.

We make enableTiming_ atomic in PGNCCL to avoid races on non-TSO hardware.

To ensure consistency, we copy its value during Work construction and replace all previous usage of enableTiming_ from the PG with usages from the Work, which now has an immutable value.





[ghstack-poisoned]

fduwjj reviewed (approved):
LGTM

kumpera committed:
Update on "[C10d] Add PG::enableCollectivesTiming to make it dynamically enabled."

Collectives timing gates the tracking when a collective starts on device.

Currently it's enabled by set the NCCL_ENABLE_TIMING env var.

The goal of this PR is to make it possible to dynamically enable that flag so users of the PG hooks don't have to set that flag in order to have their hooks work.

The design is that once set, all new collectives will have such behavior so we track it on each Work object.

We make enableTiming_ atomic in PGNCCL to avoid races on non-TSO hardware.

To ensure consistency, we copy its value during Work construction and replace all previous usage of enableTiming_ from the PG with usages from the Work, which now has an immutable value.





[ghstack-poisoned]

kumpera committed:
Update on "[C10d] Add PG::enableCollectivesTiming to make it dynamically enabled."

Collectives timing gates the tracking when a collective starts on device.

Currently it's enabled by set the NCCL_ENABLE_TIMING env var.

The goal of this PR is to make it possible to dynamically enable that flag so users of the PG hooks don't have to set that flag in order to have their hooks work.

The design is that once set, all new collectives will have such behavior so we track it on each Work object.

We make enableTiming_ atomic in PGNCCL to avoid races on non-TSO hardware.

To ensure consistency, we copy its value during Work construction and replace all previous usage of enableTiming_ from the PG with usages from the Work, which now has an immutable value.





[ghstack-poisoned]
